src/Interpreters/HashJoin.cpp

```diff
@@ -623,47 +745,50 @@ bool HashJoin::addJoinedBlock(const Block & source_block, bool check_limits)
     Block block = materializeBlock(source_block);
     size_t rows = block.rows();
 
-    ColumnRawPtrs key_columns = JoinCommon::materializeColumnsInplace(block, key_names_right);
+    size_t total_rows = 0;
+    size_t total_bytes = 0;
 
-    /// We will insert to the map only keys, where all components are not NULL.
-    ConstNullMapPtr null_map{};
-    ColumnPtr null_map_holder = extractNestedColumnsAndNullMap(key_columns, null_map);
+    // Collect all keys in all_key_names_right
+    //  and lists of indexes in this vector for all disjuncts
+    Names all_key_names_right = key_names_right.front();
+    const size_t disjuncts_num = key_names_right.size();
+    std::vector<std::vector<size_t>> key_names_right_indexes(disjuncts_num);
+    key_names_right_indexes[0].resize(all_key_names_right.size());
+    std::iota(std::begin(key_names_right_indexes[0]), std::end(key_names_right_indexes[0]), 0);
 
-    /// If RIGHT or FULL save blocks with nulls for NonJoinedBlockInputStream
-    UInt8 save_nullmap = 0;
-    if (isRightOrFull(kind) && null_map)
+    for (size_t d = 1; d < disjuncts_num; ++d)
     {
-        /// Save rows with NULL keys
-        for (size_t i = 0; !save_nullmap && i < null_map->size(); ++i)
-            save_nullmap |= (*null_map)[i];
+        for (size_t i = 0; i < key_names_right[d].size(); ++i)
+        {
+            auto it = std::find(std::cbegin(all_key_names_right), std::cend(all_key_names_right), key_names_right[d][i]);
+            if (it == std::cend(all_key_names_right))
+            {
+                key_names_right_indexes[d].push_back(all_key_names_right.size());
+                all_key_names_right.push_back(key_names_right[d][i]);
+            }
+            else
+            {
+                key_names_right_indexes[d].push_back(std::distance(std::cbegin(all_key_names_right), it));
+            }
+        }
     }
 
-    auto join_mask_col = JoinCommon::getColumnAsMask(block, condition_mask_column_name_right);
+    ColumnRawPtrs all_key_columns = JoinCommon::materializeColumnsInplace(block, all_key_names_right);
 
-    /// Save blocks that do not hold conditions in ON section
-    ColumnUInt8::MutablePtr not_joined_map = nullptr;
-    if (isRightOrFull(kind) && join_mask_col)
+    Block structured_block = structureRightBlock(block);
+    bool multiple_disjuncts = disjuncts_num > 1;
+    if (nullable_right_side && multiple_disjuncts)
     {
-        const auto & join_mask = assert_cast<const ColumnUInt8 &>(*join_mask_col).getData();
-        /// Save rows that do not hold conditions
-        not_joined_map = ColumnUInt8::create(block.rows(), 0);
-        for (size_t i = 0, sz = join_mask.size(); i < sz; ++i)
-        {
-            /// Condition hold, do not save row
-            if (join_mask[i])
-                continue;
+        JoinCommon::convertColumnsToNullable(structured_block);
+    }
```

Comment:
It is because it is not needed ;) , some sort of legacy, had been needed for the first attempt to have right joins working.
Very well spotted, thanks!